src/gen_bpf.c

```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+			(*s_head)->pri_prv = NULL;
+			(*s_head)->pri_nxt = NULL;
 		}
 	}
-	if (db_secondary != NULL) {
-		db_list_foreach(s_iter, db_secondary->syscalls) {
-			if (s_head != NULL) {
-				s_iter_b = s_head;
-				while ((s_iter_b->pri_nxt != NULL) &&
-				       (s_iter->priority <= s_iter_b->priority))
-					s_iter_b = s_iter_b->pri_nxt;
-
-				if (s_iter->priority > s_iter_b->priority) {
-					s_iter->pri_prv = s_iter_b->pri_prv;
-					s_iter->pri_nxt = s_iter_b;
-					if (s_iter_b == s_head) {
-						s_head->pri_prv = s_iter;
-						s_head = s_iter;
-					} else {
-						s_iter->pri_prv->pri_nxt =
-							s_iter;
-						s_iter->pri_nxt->pri_prv =
-							s_iter;
-					}
+}
+
+/**
+ * Sort the syscalls by syscall number
+ * @param syscalls the linked list of syscalls to be sorted
+ * @param s_head the head of the linked list to be returned to the caller
+ * @param s_tail the tail of the linked list to be returned to the caller
+ *
+ * Sorts syscalls by their syscall number
+ *
+ */
+static void _sort_syscalls_by_num(struct db_sys_list *syscalls,
+				  struct db_sys_list **s_head,
+				  struct db_sys_list **s_tail)
+{
+	struct db_sys_list *s_iter, *s_iter_b;
+
+	db_list_foreach(s_iter, syscalls) {
+		if (*s_head != NULL) {
+			s_iter_b = *s_head;
+			while ((s_iter_b->pri_nxt != NULL) &&
+			       (s_iter->num <= s_iter_b->num))
+				s_iter_b = s_iter_b->pri_nxt;
+
+			if (s_iter->num > s_iter_b->num) {
+				s_iter->pri_prv = s_iter_b->pri_prv;
+				s_iter->pri_nxt = s_iter_b;
+				if (s_iter_b == *s_head) {
+					(*s_head)->pri_prv = s_iter;
+					*s_head = s_iter;
 				} else {
-					s_iter->pri_prv = s_tail;
-					s_iter->pri_nxt = NULL;
 					s_iter->pri_prv->pri_nxt = s_iter;
-					s_tail = s_iter;
+					s_iter->pri_nxt->pri_prv = s_iter;
 				}
 			} else {
-				s_head = s_iter;
-				s_tail = s_iter;
-				s_head->pri_prv = NULL;
-				s_head->pri_nxt = NULL;
+				s_iter->pri_prv = *s_tail;
+				s_iter->pri_nxt = NULL;
+				s_iter->pri_prv->pri_nxt = s_iter;
+				*s_tail = s_iter;
 			}
+		} else {
+			*s_head = s_iter;
+			*s_tail = s_iter;
+			(*s_head)->pri_prv = NULL;
+			(*s_head)->pri_nxt = NULL;
 		}
 	}
+}
+
+/**
+ * Sort the syscalls
+ * @param syscalls the linked list of syscalls to be sorted
+ * @param s_head the head of the linked list to be returned to the caller
+ * @param s_tail the tail of the linked list to be returned to the caller
+ *
+ * Wrapper function for sorting syscalls
+ *
+ */
+static void _sort_syscalls(struct db_sys_list *syscalls,
+			   struct db_sys_list **s_head,
+			   struct db_sys_list **s_tail,
+			   unsigned int syscall_cnt)
+{
+	if (syscall_cnt < MIN_SYSCALLS_TO_USE_BINTREE)
+		_sort_syscalls_by_priority(syscalls, s_head, s_tail);
+	else
+		_sort_syscalls_by_num(syscalls, s_head, s_tail);
+}
+
+/**
+ * Count the number of syscalls in the filter
+ * @param db the filter DB
+ * @param db_secondary the secondary DB
+ *
+ * Count the total number of syscalls that are in the seccomp fitler
+ */
```

Comment:
If this information is important, it seems like we are better off keeping that count in the db_filter struct and keeping it up to date in the db.c functions, no?

Reply:
I like that idea.  Should be much more performant than recalculating it every time.

Reply:
I looked into this a little bit, and it isn't as easy as I was hoping.  For x86_64, it's pretty straightforward as the syscall count goes up by one every time a new rule is added.  But it gets tricky for x86 with its multiplexed syscalls.

It's not the most elegant, but doing the counting at bpf generation time should be foolproof and performant.  I'm afraid if we try to do it when rules are added we will see several one-off bugs in architectures with weird syscall semantics.

Reply:
We're only concerned about the number of syscalls, right?  Assuming that is the case, I think it should be as easy as having `db_rule_add()` bump a counter in the db_filter struct whenever a new syscall is added to the existing db_filter struct, yes?  The arch specific manipulations shouldn't matter since they are all going to need to call into `db_rule_add()` eventually to add a rule to the filter.

Apologies if I'm missing something very obvious :)

Reply:
Let me give it a shot again.  I played with it late one night... to very little success :(.

I'll try it again now that I have a full cup of coffee!

Reply:
Good call.  I did a quick prototype and it seems to work nicely.  Thanks.